### PR TITLE
Fix links resolution

### PR DESCRIPTION
**Pull request by ulyssessouza**
**What I did**
Provide proper link resolution to avoid inter project ambiguity

**Related issue**
Resolves https://github.com/docker/compose/issues/8935

### CODE REVIEW

File: pkg/compose/convergence.go [hunk 1]
```diff
@@ -476,6 +476,60 @@ func (s *composeService) createMobyContainer(ctx context.Context, project *types
...
+		l := linkServiceName
+		if len(linkSplit) == 2 {
+			l = linkSplit[1] // linkName if informed like in: "serviceName:linkName"
+		}
+		cnts, err := s.getContainers(ctx, projectName, oneOffExclude, true, linkServiceName)
+		if err != nil {
+			return nil
+		}
+		for _, c := range cnts {
+			containerName := strings.TrimPrefix(c.Names[0], "/")
```
Comment (ndeloof) [inline pkg/compose/convergence.go:497]: can't you use getCanonicalContainerName ?

File: pkg/compose/convergence.go [hunk 2]
```diff
@@ -476,6 +476,60 @@ func (s *composeService) createMobyContainer(ctx context.Context, project *types
...
+			links = append(links,
+				format(containerName, l),
+				format(containerName, strings.Join([]string{linkServiceName, strconv.Itoa(number)}, Separator)),
+				format(containerName, strings.Join([]string{projectName, linkServiceName, strconv.Itoa(number)}, Separator)),
+			)
+		}
+	}
+
+	if service.Labels[api.OneoffLabel] == "True" {
+		cnts, err := s.getContainers(ctx, projectName, oneOffExclude, true, service.Name)
```
Comment (ndeloof) [inline pkg/compose/convergence.go:507]: please refactor into a func to avoid code duplication

**Review by ndeloof — APPROVED** [review]
LGTM with potential minor improvements

File: pkg/compose/convergence.go (hunk 1 above)
Comment (ulyssessouza) [reply pkg/compose/convergence.go:497]: Cool!
I was looking for that function but I couldn't remember the name. 
Thanks!

File: pkg/compose/convergence.go (hunk 2 above)
Comment (ulyssessouza) [reply pkg/compose/convergence.go]: Done with a local function

**Review by ndeloof — APPROVED** [review]

File: pkg/compose/convergence.go [hunk 3]
```diff
@@ -476,6 +476,64 @@ func (s *composeService) createMobyContainer(ctx context.Context, project *types
...
+	for _, rawLink := range service.Links {
+		linkSplit := strings.Split(rawLink, ":")
+		linkServiceName := linkSplit[0]
+		l := linkServiceName
+		if len(linkSplit) == 2 {
+			l = linkSplit[1] // linkName if informed like in: "serviceName:linkName"
+		}
+		cnts, err := getServiceContainers(linkServiceName)
+		if err != nil {
+			return nil
```
Comment (mat007) [inline pkg/compose/convergence.go:498]: Is this intentional or should it be
```
		return err
```
?

File: pkg/compose/convergence.go [hunk 4]
```diff
@@ -476,6 +476,64 @@ func (s *composeService) createMobyContainer(ctx context.Context, project *types
...
+				format(containerName, strings.Join([]string{linkServiceName, strconv.Itoa(number)}, Separator)),
+				format(containerName, strings.Join([]string{projectName, linkServiceName, strconv.Itoa(number)}, Separator)),
+			)
+		}
+	}
+
+	if service.Labels[api.OneoffLabel] == "True" {
+		cnts, err := getServiceContainers(service.Name)
+		if err != nil {
+			return nil
```
Comment (mat007) [inline pkg/compose/convergence.go:513]: Same here?

File: pkg/compose/convergence.go [hunk 5]
```diff
@@ -476,6 +476,64 @@ func (s *composeService) createMobyContainer(ctx context.Context, project *types
...
+			l = linkSplit[1] // linkName if informed like in: "serviceName:linkName"
+		}
+		cnts, err := getServiceContainers(linkServiceName)
+		if err != nil {
+			return nil
+		}
+		for _, c := range cnts {
+			containerName := getCanonicalContainerName(c)
+			links = append(links,
+				format(containerName, l),
```
Comment (mat007) [inline pkg/compose/convergence.go:503]: nit: maybe `l` should be `linkName` or something a bit clearer, just like l.528?

File: pkg/compose/convergence.go (hunk 3 above)
Comment (ulyssessouza) [reply pkg/compose/convergence.go:498]: That's the right return. The return type is `[]string`

Comment (mat007) [reply pkg/compose/convergence.go:498]: But shouldn’t the return type be `([]string, error)`?

**Pull request merged**